### PR TITLE
Instructions on building feedstocks

### PR DESCRIPTION
**Pull request by dmgav**
Initial instructions on creating package recipes for Azure pipelines. Comments and corrections are welcome.

### CODE REVIEW

**Comment by dmgav** [comment]
I did some editing of the original text, now it looks somewhat more logical.

File: .gitignore [hunk 1]
```diff
@@ -10,6 +10,7 @@ __pycache__/
 .Python
 env/
 build/
+souce/_build
```
Comment (stuartcampbell) [inline .gitignore]: Looks like a typo ?

File: source/creating_packages.rst [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+--------------------
+
+Initialize feedstock using ``conda-smithy``:
+
+.. code-block:: bash
+
+    $ cd ~/packages/(your-package-name)
+    $ conda-smithy init (your-package-name)
+
+A new directory ``~packages/(your-package-name)/(your-package-name)-feedstock``
```
Comment (stuartcampbell) [inline source/creating_packages.rst]: I think ```~packages/..``` should be ```~/packages```

**Comment by dmgav** [comment]
> Would it be clearer to recommend naming the directory immediately under the `~/packages` directory to be something more generic that `(your-package-name)`?
> 
> I started to be a little confused by the
> 
> ```
> ~/packages/(your-package-name)/(your-package-name)/
> ~/packages/(your-package-name)/(your-package-name)-feedstock/
> ```
> 
> would it lose anything if it was something like:
> 
> ```
> ~/packages/nsls2forge/(your-package-name)/
> ~/packages/nsls2forge/(your-package-name)-feedstock/
> ```
> 
> I'm not saying we have to do that, just a comment - @mrakitin @jklynch what do you think ?

The directory structure is one of the confusing things. I think that it would help if I include the description of the directory structure in the text. In the example for **sixtools** package we would have the following directory structure:
`~/packages` - root directory, may have any name and placed anywhere, I just selected the convenient name.
`~/packages/event-model-feedstock` - the directory with cloned **event-model-feedstock** repository.
`~/packages/sixtools` - the directory that contains everything related to **sixtools** repository. The name doesn't matter, but it has to be package-specific.
`~/packages/sixtools/sixtools` - the directory where recipe for the new package is placed by **conda skeleton** or manually. This directory has to be named `sixtools`, since **conda-smithy** is looking for recipe in the directory with this name.
`~/packages/sixtools/sixtools-feedstock` - the directory that contains the repository with **sixtools** feedstock generated by **conda-smithy**. This repository is uploaded to `nsls-ii-forge`.

File: source/creating_packages.rst (hunk 2 above)
Comment (dmgav) [reply source/creating_packages.rst]: The typo is fixed.

**Comment by mrakitin** [comment]
We can follow the structure like this:
```
~/src/<GH org name>/<GH repo name>
```
so the example will look like
```
~/src/nsls-ii-forge/sixtools-feedstock
```
I think the initial `sixtools` directory can be removed once the feedstock is initialized as the former is a temp dir and will not be needed after that.

**Comment by dmgav** [comment]
@mrakitin Do I understand correctly, that you are proposing to recommend skipping the second level directories named after the package and put both temporary and feedstock directories in one directory named ``~/src/nsls-ii-forge/``? In this case, the directory structure for the above example of **sixtools** package will look like this:

``~/src/nsls-ii-forge/`` - root directory, may have any name and placed anywhere, I just selected the convenient name.
``~/src/nsls-ii-forge/event-model-feedstock`` - the directory with cloned event-model-feedstock repository.
``~/src/nsls-ii-forge/sixtools`` - the directory where recipe for the new package is placed by conda skeleton or manually. This directory has to be named sixtools, since conda-smithy is looking for recipe in the directory with this name. **This directory is temporary and may be deleted after the feedstock is initialized.**
``~/src/nsls-ii-forge/sixtools-feedstock`` - the directory that contains the repository with sixtools feedstock generated by conda-smithy. This repository is uploaded to nsls-ii-forge.

File: source/creating_packages.rst [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+
+.. code-block::
+
+  doc_url: ""
+  dev_url: ""
+
+Additional steps:
+
+  -- Add license file name ``LICENSE`` to ``about`` section.
+  -- Update URL.
```
Comment (jklynch) [inline source/creating_packages.rst]: -- Update ``home`` URL in ``about`` section to point to the package's GitHub repository.

File: source/creating_packages.rst [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+
+    TODO: Some notes on editing ``meta.yaml`` file.
+
+Prepare Recipe Files
+====================
+
+Collect additional files
+------------------------
+
+Copy ``conda_build_config.yaml`` files into your recipe directory:
```
Comment (jklynch) [inline source/creating_packages.rst]: files -> file ?

File: source/creating_packages.rst [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+.. code-block::
+
+    channel_sources:
+    - nsls2forge,defaults
+    channel_targets:
+    - nsls2forge main
+    python:
+    - '3.6'
+
+If you are building noarch package, then close the file. For arch packages with
```
Comment (jklynch) [inline source/creating_packages.rst]: ... building a noarch package ...

File: source/creating_packages.rst [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+
+Replace ``conda-forge.yml`` in the feedstock directory with ``conda-forge.yml`` from
+``event-model-feedstock`` package:
+
+.. code-block:: bash
+
+    $ cd ~/packages
+    $ cp event-model-feedstock/conda-forge.yml (your-package-name)/(your-package-name)-feedstock/
+
+Define Asure variables
```
Comment (jklynch) [inline source/creating_packages.rst]: Asure -> Azure

File: source/creating_packages.rst [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+
+Create pull request at GitHub
+=============================
+
+Open github page ``https://github.com/nsls-ii-forge/(your-package-name)-feedstock``
+and create pull request. In pull request comments include a brief note and **the link to the original 
+repository** of the package (PyPI, conda-forge or GitHub).
+
+Closely examine build results to ensure that the packages were built for all systems
+and Python version and all tests passed successfully. Correct issues if necessary.
```
Comment (jklynch) [inline source/creating_packages.rst]: and Python versions

File: source/creating_packages.rst [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+Each time the change is made to configuration files, the feedstock must be
```
Comment (jklynch) [inline source/creating_packages.rst]: the change -> a change

File: source/creating_packages.rst [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+rerendered and changes must be committed and pushed. Merge the pull request once
+all issues are fixed.
+
+Associate Anaconda Token from Variable Groups With the New Pipeline
+===================================================================
+
+- Log into `dev.azure.com <https://dev.azure.com>`_. 
+- Select the pipeline named ``(your-package-name)-feedstock``. 
+- Click ``Edit``.
+- Click a button with three vertical dots in the right top corner.
```
Comment (jklynch) [inline source/creating_packages.rst]: a button -> the button

**Comment by dmgav** [comment]
@jklynch  Josh, thank you for comments. I corrected all mentioned typos and errors. We would need @mrakitin to read it. If he agrees with the outline of the procedure, we can probably merge it and submit further corrections as pull requests.

**Comment by dmgav** [comment]
The new corrected version of instructions is uploaded.

**Comment by dmgav** [comment]
I am not sure, why the suggestions were not applied. I went through all suggestions. I went through the list manually again and I found that two suggestions were not applied, so they should be applied now.

**Comment by mrakitin** [comment]
Weird, I am still seeing quite a few suggestions at https://github.com/nsls-ii-forge/docs/pull/2/files#diff-6c3e18038b6a7d4113c7c4d7c4446e3c. Do you see any? If not, may be a GH glitch...

**Comment by dmgav** [comment]
I can see changes if I follow the link. Ok, I will go through them manually and upload the new file.

File: source/creating_packages.rst [hunk 10]
```diff
@@ -0,0 +1,405 @@
...
+
+Open github page ``https://github.com/nsls-ii-forge/(package-name)-feedstock``
+and create pull request. In pull request comments include a brief note and **the link to the original 
+repository** of the package (PyPI, conda-forge or GitHub).
+
+Closely examine build results to ensure that the packages were built for all systems
+and Python versions and all tests passed successfully. Correct issues if necessary.
+Each time a change is made to configuration files, the feedstock must be
+rerendered and changes must be committed and pushed. Merge the pull request once
+all issues are fixed.
```
[the remark this answers is not in the document]
Comment (dmgav) [reply source/creating_packages.rst]: Fixed

File: source/creating_packages.rst [hunk 11]
```diff
@@ -0,0 +1,405 @@
...
+    $ cp event-model-feedstock/conda-forge.yml (package-name)-feedstock/
+
+Define Azure variables
+======================
+
+.. code-block:: bash
+
+    $ export AZURE_ORG_OR_USER=nsls2forge
+    $ export AZURE_PROJECT_NAME=nsls2forge
+
```
[the remark this answers is not in the document]
Comment (dmgav) [reply source/creating_packages.rst]: The note have been added.

File: source/creating_packages.rst [hunk 12]
```diff
@@ -0,0 +1,405 @@
...
+.. code-block:: bash
+
+    $ cd ~/src/nsls-ii-forge
+    $ mkdir (package-name)
+    $ cd (package-name)
+
+The recipe may be created based on ``meta.yaml`` file from
+the original package repository. If such file is not available
+or unusable, find an appropriate sample ``meta.yaml`` (from
```
[the remark this answers is not in the document]
Comment (dmgav) [reply source/creating_packages.rst]: Fixed

File: source/creating_packages.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (mrakitin) [reply source/creating_packages.rst]: Great!

**Comment by mrakitin** [comment]
@jklynch, I will merge this PR as all your comments seem to be applied. The follow-up updates can be done via separate PRs.

**Pull request merged**